frontend/src/components/Header/StopClock.js

```diff
@@ -0,0 +1,101 @@
+// @flow
+import React, { Component } from 'react';
+import CircularProgressbar from 'react-circular-progressbar';
+import 'react-circular-progressbar/dist/styles.css';
+
+type Props = {
+  startTime: number, // Date.now() ms at start
+  countdown: number, // delta in seconds
+  currentTime: number, // Date.now() at present
+  history: *,
+};
+
+type State = {
+  currentTime: number,
+};
+
+type CircularProgressbarProps = {
+  text: string,
+  percentage: number,
+};
+
+const StyledCircularProgressbar = (props: CircularProgressbarProps) => (
+  <CircularProgressbar
+    percentage={props.percentage}
+    text={props.text}
+    strokeWidth={10}
+    styles={{
+      root: {},
+      path: {
+        stroke: '#1c8bf4',
+        strokeLinecap: 'butt',
+        transition: 'stroke-dashoffset 0.5s ease 0s',
+      },
+      trail: {
+        stroke: '#e6e6e6',
+      },
+      text: {
+        fill: '#1c8bf4',
+        fontSize: '30px',
+      },
+    }}
+  />
+);
+
+const getPercRemaining = (
+  currentTime: number,
+  startTime: number,
+  countdown: number
+) => {
+  return (currentTime - startTime) / (countdown * 1000);
+};
+
+const getTimeLabel = (secondsRemaining: number) =>
+  secondsRemaining < 60
+    ? `${Math.ceil(secondsRemaining)}s`
+    : `${Math.ceil(secondsRemaining / 60)}m`;
+
+class StopClock extends Component<Props, State> {
+  state: State = {
+    currentTime: Date.now(),
+  };
+
+  tick = () => {
+    this.setState({
+      currentTime: Date.now(),
+    });
+  };
+
+  timer = null;
+  endTime = 0;
+  componentDidMount() {
+    this.timer = setInterval(this.tick, 500);
+    this.endTime = this.props.startTime + 1000 * this.props.countdown;
+  }
+
+  componentDidUpdate() {
+    if (this.endTime !== 0 && Date.now() >= this.endTime) {
+      clearInterval(this.timer);
+      this.endTime = 0;
+      this.props.history.push('/afterward');
```

Comment:
Having the logic for redirecting a user after the exercise is finished in the clock component doesn't feel like the best place for it. You might want to do other things when the exercise is completed (like tell the server) so this bit of code could get quite complicated. I think the logic should be in the `exercise` action `tickTimer` where you have access to the length of the exercise and also the current time. What do you think?

Reply:
I agree it shouldn't be in here, but I think the best thing to do is pass the onTimeout function into props and just call it when the clock times out

Reply:
changes here: https://github.com/cybsafe/phishtray/pull/91/commits/a53a752aa090f132acb2a7e4cc2a79f17cc30107